File: core/utils/filter.go
package utils

import (
	"github.com/crawlab-team/crawlab/core/constants"
	"github.com/crawlab-team/crawlab/core/interfaces"
	"go.mongodb.org/mongo-driver/bson"
)

// FilterToQuery Translate entity.Filter to bson.M
func FilterToQuery(f interfaces.Filter) (q bson.M) {
	if f == nil || f.IsNil() {
		return nil
	}

	q = bson.M{}
	for _, cond := range f.GetConditions() {
		key := cond.GetKey()
		op := cond.GetOp()
		value := cond.GetValue()
		switch op {
		case constants.FilterOpNotSet:
			// do nothing
		case constants.FilterOpEqual:
			q[key] = cond.GetValue()
		case constants.FilterOpNotEqual:
			q[key] = bson.M{"$ne": value}
		case constants.FilterOpContains, constants.FilterOpContainsShort, constants.FilterOpRegex, constants.FilterOpSearch, constants.FilterOpSearchShort:
			q[key] = bson.M{"$regex": value, "$options": "i"}
		case constants.FilterOpNotContains, constants.FilterOpNotContainsShort:
			q[key] = bson.M{"$not": bson.M{"$regex": value}}
		case constants.FilterOpIn:
			q[key] = bson.M{"$in": value}
		case constants.FilterOpNotIn:
			q[key] = bson.M{"$nin": value}
		case constants.FilterOpGreaterThan:
			q[key] = bson.M{"$gt": value}
		case constants.FilterOpGreaterThanEqual:
			q[key] = bson.M{"$gte": value}
		case constants.FilterOpLessThan:
			q[key] = bson.M{"$lt": value}
		case constants.FilterOpLessThanEqual:
			q[key] = bson.M{"$lte": value}
		default:
			// ignore invalid operation
		}
	}
	if len(q) == 0 {
		return nil
	}
	return q
}

File: frontend/crawlab-ui/src/components/ui/filter/filter.ts
import { plainClone } from '@/utils';
import {
  FILTER_OP_CONTAINS,
  FILTER_OP_EQUAL,
  FILTER_OP_GREATER_THAN,
  FILTER_OP_GREATER_THAN_EQUAL,
  FILTER_OP_LESS_THAN,
  FILTER_OP_LESS_THAN_EQUAL,
  FILTER_OP_NOT_CONTAINS,
  FILTER_OP_NOT_EQUAL,
  FILTER_OP_NOT_SET,
  FILTER_OP_REGEX,
} from '@/constants';

export const defaultFilterCondition: FilterConditionData = {
  op: FILTER_OP_NOT_SET,
  value: '',
};

export const getDefaultFilterCondition = () => {
  return plainClone(defaultFilterCondition);
};

export const conditionTypesOptions: SelectOption[] = [
  { value: FILTER_OP_NOT_SET, label: 'Not Set' },
  { value: FILTER_OP_CONTAINS, label: 'Contains' },
  { value: FILTER_OP_NOT_CONTAINS, label: 'Not Contains' },
  { value: FILTER_OP_REGEX, label: 'Regex' },
  { value: FILTER_OP_EQUAL, label: 'Equal to' },
  { value: FILTER_OP_NOT_EQUAL, label: 'Not Equal to' },
  { value: FILTER_OP_GREATER_THAN, label: 'Greater than' },
  { value: FILTER_OP_LESS_THAN, label: 'Less than' },
  { value: FILTER_OP_GREATER_THAN_EQUAL, label: 'Greater than or Equal to' },
  { value: FILTER_OP_LESS_THAN_EQUAL, label: 'Less than or Equal to' },
];

export const conditionTypesMap: { [key: string]: string } = (() => {
  const map: { [key: string]: string } = {};
  conditionTypesOptions.forEach(d => {
    map[d.value] = d.label as string;
  });
  return map;
})();
